.github/workflows/release.yml

```diff
@@ -0,0 +1,267 @@
+
+    - name: Build via Maven and run tests
+      env:
+        MAVEN_GPG_PASSPHRASE: ${{ secrets.MAVEN_GPG_PASSPHRASE }}
+        SPARK_LOCAL_IP: localhost
+      run: |
+        ./mvnw install -Pnative -Pjdk8-tests -Pcode-coverage -Prelease
+
+    - name: Tag the Docker image
+      run: |
+        # We can always use projectnessie/nessie here because it's a release
+        docker tag projectnessie/nessie:${RELEASE_VERSION} projectnessie/nessie:latest
+
+    - name: Build Gradle plugin and run tests
+      run: |
+        (cd tools/apprunner-gradle-plugin/ ; ./gradlew build)
+
+    - name: Build pynessie and run tests
+      run: |
+        cd python
+        tox
+        python setup.py sdist bdist_wheel
+
+    # TODO Remove this one - b/c it's codecov?
+    - uses: codecov/codecov-action@v1
+      with:
+        flags: java,python
+
+    # TODO Seems we don't capture the Python test results
+    # TODO Seems we don't capture the Gradle test results
+    - name: Capture test results
+      uses: actions/upload-artifact@v2
+      with:
+        name: test-results
+        path: |
+          **/target/surefire-reports/*
+          **/target/failsafe-reports/*
+
+    # Record the release-version in git and add the git tag for the release.
+    - name: Record ${{ github.event.inputs.releaseVersion }} release in git
+      env:
+        RELEASE_VERSION: ${{ github.event.inputs.releaseVersion }}
+        GIT_TAG: nessie-${{ github.event.inputs.releaseVersion }}
+      run: |
+        git commit -a -m "[release] release nessie-${RELEASE_VERSION}"
+        git tag ${GIT_TAG}
+
+    # Update versions to next development iteration and add a new git commit
+    - name: Set next development version version ${{ github.event.inputs.nextVersion }}
+      env:
+        NEXT_VERSION: ${{ github.event.inputs.nextVersion }}
+      run: |
+        ./mvnw versions:set -DgenerateBackupPoms=false -DnewVersion=${NEXT_VERSION}-SNAPSHOT
+        (cd python ; bumpversion --no-commit --new-version ${NEXT_VERSION} minor)
+
+        git commit -a -m "[release] next development iteration ${NEXT_VERSION}-SNAPSHOT"
+
+    # Push the 2 git commits and git tag. If this one fails, some other commit was pushed to the
+    # 'main' branch and break the linear history for the Nessie git repo.
+    - name: Push tag + branch
+      env:
+        GIT_TAG: nessie-${{ github.event.inputs.releaseVersion }}
+      run: |
+        UPSTREAM="$(git remote)"
+
+        git push ${UPSTREAM}
+        git push ${UPSTREAM} :refs/tags/${GIT_TAG}
+        git push ${UPSTREAM} ${GIT_TAG}
+
+    # This last step finally deploys the artifacts. All build and test steps were already ran
+    # in previous steps
+    # TODO close + deploy the OSSRH staging repo, maybe switch from maven-deploy-plugin to nexus-staging-maven-plugin to allow that?
+    - name: Publish artifacts for release ${{ github.event.inputs.releaseVersion }}
```

Comment:
I would rather do this as separate steps. And re-use actions where possible (eg pypa has a great pypi action)

Reply:
See next commit - splitting the "build + test" from "deployment"